USWDS: style aria-disabled to match disabled

amyleadem opened the pull request:
<!-- Please feel free to remove whatever sections/lines in this aren’t relevant.

Use the title line as the title of your pull request, then delete these lines.

## Title line template: [Title]: Brief description

Website: For pull requests that impact designsystem.digital.gov’s look, feel, or functionality, please open a pull request on the uswds-site repo (https://github.com/uswds/uswds-site).

-->

## Description

Closes: https://github.com/uswds/uswds/issues/4508

Preview links:
- [Button](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-aria-disabled/?path=/story/components-button--default)
- [Button Group](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-aria-disabled/?path=/story/components-button-group--default)
- [Checkbox](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-aria-disabled/?path=/story/components-form-inputs-checkbox--default)
- [Date Picker](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-aria-disabled/?path=/story/components-form-inputs-date-picker--aria-disabled)
- [Radio](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cloud.gov/preview/uswds/uswds/al-aria-disabled/?path=/story/components-form-inputs-radio--default)

### Problem
At times, users of the Design System will want to use `aria-disabled` in place of the `disabled` attribute, but are limited because currently there is no explicit styling for `aria-disabled` elements in the Design System. 

`aria-disabled` provides a different experience to end-users that use keyboard navigation or assistive technologies. The primary difference between the two attributes is this:
-  `disabled` elements _do not_ receive focus on tab, which can result in the item being:
    - invisible to users on screen readers who use tab to navigate forms
    - difficult and confusing to interact with for other users using keyboard navigation without a screen reader, particularly for those with cognitive and fine-motor disabilities 
- `aria-disabled` _does_ receive focus on tab, which can give users important clues about their next steps

The movement in industry conversation seems to be towards `aria-disabled` as the preferred user experience  (See resources below), but other mainstream design systems still use the `disabled` attribute. This item is only one piece of a larger discussion about the UX/accessibility of disabled buttons. 

From [Mozilla](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-disabled):
> The aria-disabled attribute is similar to the HTML disabled attribute, but better. The first rule of ARIA is "If you can use a native HTML element or attribute with the semantics and behavior you require already built in, instead of repurposing an element and adding an ARIA role, state or property to make it accessible, then do so." aria-disabled is an exception to that rule. It is one of the only cases where an ARIA attribute may be better than the native disabled attribute HTML equivalent.

#### Resources
- [Making disabled buttons more inclusive](https://css-tricks.com/making-disabled-buttons-more-inclusive/) (CSS Tricks)
- [aria-disabled](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-disabled) (Mozilla)
- [Videos comparing experience of disabled and aria-disabled](https://a11y-101.com/development/aria-disabled) (A11y-101)
### Solution
Providing styling for both `aria-disabled` and `disabled` will give users more flexibility to make the appropriate usability decisions for their site. 

:warning: This PR is concerned only with CSS styling of `aria-disabled`. It is not intended to replace `disabled` as the default attribute in our components.

### Tasks completed
- [x] Add styling for aria-disabled wherever `:disabled` is declared
- [x] Standardize hover states for disabled/aria-disabled buttons 
    - Note: This likely addresses the concerns in: https://github.com/uswds/uswds/issues/4420
- [x] Set disabled/aria-disabled cursor to `auto` instead of `pointer`
    - Note: Radio and Checkbox do use `cursor: not-allowed`. Should we standardize to one value?
- [x] Add aria-disabled control to relevant stories to help with testing

### Testing
1. Confirm that all `:disabled` SCSS references also include `aria-disabled`
1. Check that the styling for `disabled` and `aria-disabled` align on the following components:
    - Button
    - Button group
    - Date picker 
    - Radio
    - Checkbox
    
## Additional information

Before you hit Submit, make sure you’ve done whichever of these applies to you:

- [x] Follow the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `npm test` and make sure the tests for the files you have changed have passed.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.
- [x] Title your pull request using this format: [Website] - [UI component]: Brief statement describing what this pull request solves.

amyleadem commented on packages/usa-button-group/src/content/usa-button-group.json:
```diff
@@ -3,5 +3,6 @@
     "<button class=\"usa-button\">Continue</button>"
-  ]
+  ],
+  "show_disabled": false
```
That works for me. Updated in [855072b](https://github.com/uswds/uswds/pull/4783/commits/855072b201e64649e993df7c31cd3416c598c7db)

amyleadem commented on packages/usa-button-group/src/styles/_usa-button-group.scss:
```diff
@@ -124,10 +124,15 @@
-      border-right-color: color("base");
+    &:not(:last-child) .usa-button--accent-cool[aria-disabled="true"]::before,
+    &:not(:last-child) .usa-button--base:disabled::before,
+    &:not(:last-child) .usa-button--base[aria-disabled="true"]::before {
```
Good point. I have added `:where` to the `:not` declarations in this file and all seems to be working well. Thanks for the suggestion to clean this up!
Updates found in [75643d7](https://github.com/uswds/uswds/pull/4783/commits/75643d7db67449c4e306c93f389fa380c5d88400)

mejiaj reviewed (changes requested):
Looks good so far! Added some comments below. Show disabled is for `aria-disabled` exclusively, right? Do you think it'd be easier to add that attribute on the existing elements vs adding new items?

amyleadem commented on packages/usa-button-group/src/content/usa-button-group~segmented.json:
```diff
@@ -4,5 +4,6 @@
     "<button class=\"usa-button usa-button--outline\">Hybrid</button>"
-  ]
+  ],
+  "show_disabled": true
```
Hmmm...it is working well for me both locally and on Federalist. For good measure I checked on all my installed browsers as well. Would you mind checking again?

amyleadem commented on packages/usa-button-group/src/styles/_usa-button-group.scss:
```diff
@@ -93,52 +93,47 @@
+      [class*="usa-button"]:disabled::before,
+      [class*="usa-button"][aria-disabled="true"]::before {
+        border-right-color: color($theme-body-background-color);
+      }
```
Good catch. Updated in [beecb2f](https://github.com/uswds/uswds/pull/4783/commits/beecb2f881e02c2e3ea9e2a06d9647e4b26ed0ad).

amyleadem commented on packages/usa-checkbox/src/usa-checkbox.twig:
```diff
@@ -16,4 +16,10 @@
+      <input class="usa-checkbox__input" id="check-historical-carver" type="checkbox" name="historical-figures" value="george-washington-carver" disabled>
+      <label class="usa-checkbox__label" for="check-historical-carver">George Washington Carver</label>
+    </div>
+  {% endif %}
```
I dig it. Updated in [de783f3](https://github.com/uswds/uswds/pull/4783/commits/de783f3fe9c47336eb1dd2a4b948557151195a44).

amyleadem replied:
Also added this behavior to usa-radio: [acd1289](https://github.com/uswds/uswds/pull/4783/commits/acd128973c1f847572f2e0a8c64f3092cac1a3dd)

amyleadem replied:
And added to button group: [01dafee](https://github.com/uswds/uswds/pull/4783/commits/01dafee042d688fac29c43c56a2473f2be307ec9)

amyleadem commented:
@mejiaj 
Thanks for the comments. They were helpful for reframing approach. Here are the changes I made:
- Updated the structure of the button-group segemented story to display all button variants [01dafee](https://github.com/uswds/uswds/pull/4783/commits/01dafee042d688fac29c43c56a2473f2be307ec9)
- Added `aria-disabled` to the element rather than creating new elements in:
    - Button group [01dafee](https://github.com/uswds/uswds/pull/4783/commits/01dafee042d688fac29c43c56a2473f2be307ec9)
    - Radio [acd1289](https://github.com/uswds/uswds/pull/4783/commits/acd128973c1f847572f2e0a8c64f3092cac1a3dd)
    - Checkbox [de783f3](https://github.com/uswds/uswds/pull/4783/commits/de783f3fe9c47336eb1dd2a4b948557151195a44)
    - I did not do this for button because it would require creating some demo styles for `.usa-button--active` and `.usa-button--hover`
- Repaired date picker stories [ea3389e](https://github.com/uswds/uswds/pull/4783/commits/ea3389ea275e2a5a4e11b1e6014c42d1558ab995)

Please let me know if you have any questions or concerns.

amyleadem commented:
> Sorry, one last item. Do you think the disabled controls might be better if we combined them into radio or select? Because right now you can have both disabled and aria-disabled selected.

Good call! I'll fix now.

mejiaj commented on packages/usa-button-group/src/usa-button-group.twig:
```diff
@@ -1,7 +1,47 @@
+  {% else %}
+  <ul class="usa-button-group">
+    <li class="usa-button-group__item">
+      <a href="javascript:void(0);" class="usa-button usa-button--outline" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Back</a>
```
Do you think we should remove `disabled` from these link buttons if `disabled` isn't a valid attribute for them?

amyleadem replied:
Good point. Removed in [af3c60e](https://github.com/uswds/uswds/pull/4783/commits/af3c60ef97ce4bcc4d7676dd447901cc0f1a3fb5)

mejiaj commented on packages/usa-button-group/src/usa-button-group.twig:
```diff
@@ -1,7 +1,47 @@
+  <ul class="usa-button-group usa-button-group--segmented">
     <li class="usa-button-group__item">
-        {{ item }}
+      <button class="usa-button" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Map</button>
```
Maybe an if/elseif statement might be more readable  —  along with some formatting.

```suggestion
      <button class="usa-button"
        {% if disabled_state == "disabled" %}
          disabled
        {% elseif disabled_state == "aria-disabled"  %}
          aria-disabled="true"
        {% endif %}
      >
        Map
      </button>
```

amyleadem replied:
Opened issue #4881 to resolve in a future enhancement

mejiaj commented on packages/usa-button-group/src/usa-button-group.twig:
```diff
@@ -1,7 +1,47 @@
+{% if segmented == true %}
+  <ul class="usa-button-group usa-button-group--segmented">
     <li class="usa-button-group__item">
-        {{ item }}
+      <button class="usa-button" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Map</button>
     </li>
+    <li class="usa-button-group__item">
+      <button class="usa-button usa-button--outline" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Hybrid</button>
+    </li>
+    <li class="usa-button-group__item">
+      <button class="usa-button usa-button--outline" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Satellite</button>
+    </li>
```
We could shorten this with a loop because we have an array of items.

```twig
  <ul class="usa-button-group usa-button-group--segmented">
    {% for item in items %}
      <li class="usa-button-group__item">
        <button class="usa-button"
          {% if disabled_state == "disabled" %}
            disabled
          {% elseif disabled_state == "aria-disabled"  %}
            aria-disabled="true"
          {% endif %}
        >
          {{ item }}
        </button>
      </li>
    {% endfor %}
  </ul>
```

amyleadem replied:
Opened issue #4881 to resolve in a future enhancement

mejiaj commented on packages/usa-button-group/src/usa-button-group.twig:
```diff
@@ -1,7 +1,47 @@
+    </li>
+  </ul>
+
+  {% for modifier in modifiers %}
+  {% if 'usa-button--outline usa-button--inverse' in modifier %}
+    <div class="bg-base-darkest padding-1" style="max-width: fit-content">
+  {% endif %}
+    <ul class="usa-button-group usa-button-group--segmented">
+      <li class="usa-button-group__item">
+        <button class="usa-button {{ modifier }}" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Map</button>
+      </li>
+      <li class="usa-button-group__item">
+        <button class="usa-button {{ modifier }}" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Hybrid</button>
+      </li>
+      <li class="usa-button-group__item">
+        <button class="usa-button {{ modifier }}" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Satellite</button>
+      </li>
+    </ul>
+  {% if 'usa-button--outline usa-button--inverse' in modifier %}
+    </div>
+  {% endif %}
```
If we added items to `usa-button-group~segmented.json` we could loop too.

amyleadem replied:
Opened issue #4881 to resolve in a future enhancement

mejiaj commented on packages/usa-button-group/src/usa-button-group.twig:
```diff
@@ -1,7 +1,47 @@
+<!-- end segmented styles -->
+
+<!-- start default styles -->
+  {% else %}
+  <ul class="usa-button-group">
+    <li class="usa-button-group__item">
+      <a href="javascript:void(0);" class="usa-button usa-button--outline" {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Back</a>
+    </li>
+    <li class="usa-button-group__item">
+        <button class="usa-button" {% if disabled_state == "disabled" %}disabled{% endif %} {% if disabled_state == "aria-disabled" %}aria-disabled="true"{% endif %}>Continue</button>
+    </li>
+  </ul>
```
Not sure what this does, but it seems like this template is trying to do three different things. We should clarify where possible.

amyleadem replied:
Opened issue https://github.com/uswds/uswds/issues/4881 to resolve in a future enhancement

mejiaj replied:
Great, thanks @amyleadem! Resolving this thread.

mejiaj reviewed (changes requested):
Added some suggestions and comments for clarifying button group template. I realize this is outside the scope of the PR, so happy to move it to a separate issue. If you think so too, can you create a new issue for it?

The styling looks good.

amyleadem commented:
Note: Opened issue #4881 to improve the usability of the button-group template in a future enhancement.
